File: PYTHON/vehiclemodels/sim/low_speed_safety.py
"""Low-speed stabilisation utilities for vehicle simulations."""
from __future__ import annotations

import math
from dataclasses import asdict, dataclass
from typing import MutableSequence, Optional, Sequence


@dataclass(frozen=True)
class LowSpeedSafetyProfile:
    engage_speed: float
    release_speed: float
    yaw_rate_limit: float
    slip_angle_limit: float

    def __post_init__(self) -> None:  # type: ignore[override]
        if self.engage_speed < 0.0:
            raise ValueError("engage_speed must be non-negative")
        if self.release_speed <= 0.0:
            raise ValueError("release_speed must be positive")
        if self.release_speed < self.engage_speed:
            raise ValueError("release_speed must be >= engage_speed")
        if self.yaw_rate_limit <= 0.0:
            raise ValueError("yaw_rate_limit must be positive")
        if self.slip_angle_limit <= 0.0:
            raise ValueError("slip_angle_limit must be positive")


@dataclass(frozen=True)
class LowSpeedSafetyConfig:
    """Configuration thresholds for engaging low-speed safeguards."""

    normal: LowSpeedSafetyProfile
    drift: LowSpeedSafetyProfile
    stop_speed_epsilon: float
    drift_enabled: bool = False

    def __post_init__(self) -> None:  # type: ignore[override]
        if isinstance(self.normal, dict):
            object.__setattr__(self, "normal", LowSpeedSafetyProfile(**self.normal))
        if isinstance(self.drift, dict):
            object.__setattr__(self, "drift", LowSpeedSafetyProfile(**self.drift))
        self.normal.__class__(**asdict(self.normal))
        self.drift.__class__(**asdict(self.drift))
        if self.stop_speed_epsilon < 0.0:
            raise ValueError("stop_speed_epsilon must be non-negative")


class LowSpeedSafety:
    """State limiter to keep vehicle models well-behaved at very low speeds."""

    def __init__(
        self,
        config: LowSpeedSafetyConfig,
        *,
        longitudinal_index: Optional[int],
        lateral_index: Optional[int] = None,
        yaw_rate_index: Optional[int],
        slip_index: Optional[int],
        wheel_speed_indices: Optional[Sequence[int]] = None,
        steering_index: Optional[int] = None,
        wheelbase: Optional[float] = None,
        rear_length: Optional[float] = None,
    ) -> None:
        self.config = config
        self._longitudinal_index = longitudinal_index
        self._lateral_index = lateral_index
        self._yaw_rate_index = yaw_rate_index
        self._slip_index = slip_index
        self._wheel_speed_indices: tuple[int, ...] = (
            tuple(int(idx) for idx in wheel_speed_indices)
            if wheel_speed_indices
            else ()
        )
        self._steering_index = steering_index
        self._wheelbase = float(wheelbase) if wheelbase is not None else None
        self._rear_length = float(rear_length) if rear_length is not None else None
        self._engaged = False
        self._drift_enabled = bool(config.drift_enabled)

    @property
    def engaged(self) -> bool:
        return self._engaged

    @property
    def drift_enabled(self) -> bool:
        return self._drift_enabled

    def set_drift_enabled(self, enabled: bool) -> None:
        self._drift_enabled = bool(enabled)

    def reset(self) -> None:
        self._engaged = False

    def apply(
        self,
        state: MutableSequence[float],
        speed: float,
        *,
        update_latch: bool = True,
    ) -> None:
        """Clamp unstable states when operating near standstill."""

        profile = self.config.drift if self._drift_enabled else self.config.normal
        stop_eps = self.config.stop_speed_epsilon

        if update_latch:
            if self._engaged:
                if speed > profile.release_speed:
                    self._engaged = False
            else:
                if speed < profile.engage_speed:
                    self._engaged = True

        yaw_target = self._kinematic_yaw_rate(state, speed)
        slip_target = self._kinematic_slip(state, speed)
        lateral_target = self._kinematic_lateral_velocity(state, speed)

        if self._engaged and speed <= stop_eps:
            beta_ref = self._velocity_slip(state)
            beta_ref = 0.0 if beta_ref is None else beta_ref
            yaw_target = 0.0
            slip_target = beta_ref
            lateral_target = speed * math.sin(beta_ref)

        if self._longitudinal_index is not None:
            idx = self._longitudinal_index
            value = float(state[idx])
            if value < 0.0:
                state[idx] = 0.0
            elif abs(value) <= stop_eps and not self._engaged:
                state[idx] = 0.0

        if self._yaw_rate_index is not None:
            idx = self._yaw_rate_index
            if self._engaged:
                limit = profile.yaw_rate_limit
                target = 0.0 if yaw_target is None else yaw_target
                state[idx] = max(-limit, min(limit, target))
            else:
                limit = profile.yaw_rate_limit
                state[idx] = max(-limit, min(limit, float(state[idx])))

        if self._lateral_index is not None:
            idx = self._lateral_index
            value = float(state[idx])
            if self._engaged:
                if lateral_target is None:
                    limit = stop_eps
                    state[idx] = max(-limit, min(limit, value))
                else:
                    state[idx] = lateral_target
            elif abs(value) <= stop_eps:
                state[idx] = 0.0

        if self._slip_index is not None:
            idx = self._slip_index
            if self._engaged:
                limit = profile.slip_angle_limit
                target = 0.0 if slip_target is None else slip_target
                state[idx] = max(-limit, min(limit, target))
            else:
                limit = profile.slip_angle_limit
                state[idx] = max(-limit, min(limit, float(state[idx])))

        if self._wheel_speed_indices:
            wheel_stage_latch = self._engaged or (
                not update_latch and speed < profile.engage_speed
            )
            for idx in self._wheel_speed_indices:
                value = float(state[idx])
                if value <= 0.0:
                    state[idx] = 0.0
                elif wheel_stage_latch and value <= stop_eps:
                    state[idx] = 0.0


    def _kinematic_beta(self, state: Sequence[float]) -> Optional[float]:
        if (
            self._steering_index is None
            or self._wheelbase is None
            or self._rear_length is None
            or self._wheelbase <= 0.0
        ):
            return None
        delta = float(state[self._steering_index])
        ratio = self._rear_length / self._wheelbase
        return math.atan(math.tan(delta) * ratio)

    def _kinematic_yaw_rate(
        self, state: Sequence[float], speed: float
    ) -> Optional[float]:
        beta = self._kinematic_beta(state)
        if beta is None or self._wheelbase is None or self._steering_index is None:
            return None
        delta = float(state[self._steering_index])
        if abs(speed) <= 1e-9:
            return 0.0
        return speed * math.cos(beta) * math.tan(delta) / self._wheelbase

    def _kinematic_slip(
        self, state: Sequence[float], _speed: float
    ) -> Optional[float]:
        return self._kinematic_beta(state)

    def _kinematic_lateral_velocity(
        self, state: Sequence[float], speed: float
    ) -> Optional[float]:
        beta = self._kinematic_beta(state)
        if beta is None:
            return None
        return speed * math.sin(beta)

    def _velocity_slip(self, state: Sequence[float]) -> Optional[float]:
        if self._longitudinal_index is None or self._lateral_index is None:
            return None
        long_idx = self._longitudinal_index
        lat_idx = self._lateral_index
        if long_idx >= len(state) or lat_idx >= len(state):
            return None

        longitudinal = float(state[long_idx])
        lateral = float(state[lat_idx])
        if abs(longitudinal) <= 1e-9 and abs(lateral) <= 1e-9:
            return 0.0
        return math.atan2(lateral, longitudinal)


__all__ = ["LowSpeedSafetyProfile", "LowSpeedSafetyConfig", "LowSpeedSafety"]

File: src/sim/low_speed_safety.cpp
#include "sim/low_speed_safety.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace vehiclemodels::sim {

void LowSpeedSafetyConfig::validate() const
{
    if (engage_speed < 0.0) {
        throw std::invalid_argument("engage_speed must be non-negative");
    }
    if (release_speed <= 0.0) {
        throw std::invalid_argument("release_speed must be positive");
    }
    if (release_speed < engage_speed) {
        throw std::invalid_argument("release_speed must be >= engage_speed");
    }
    if (yaw_rate_limit <= 0.0) {
        throw std::invalid_argument("yaw_rate_limit must be positive");
    }
    if (slip_angle_limit <= 0.0) {
        throw std::invalid_argument("slip_angle_limit must be positive");
    }
    if (stop_speed_epsilon < 0.0) {
        throw std::invalid_argument("stop_speed_epsilon must be non-negative");
    }
}

LowSpeedSafety::LowSpeedSafety(const LowSpeedSafetyConfig& config,
                               std::optional<int> longitudinal_index,
                               std::optional<int> lateral_index,
                               std::optional<int> yaw_rate_index,
                               std::optional<int> slip_index,
                               std::vector<int> wheel_speed_indices,
                               std::optional<int> steering_index,
                               std::optional<double> wheelbase,
                               std::optional<double> rear_length)
    : config_(config)
    , longitudinal_index_(longitudinal_index)
    , lateral_index_(lateral_index)
    , yaw_rate_index_(yaw_rate_index)
    , slip_index_(slip_index)
    , wheel_speed_indices_(std::move(wheel_speed_indices))
    , steering_index_(steering_index)
    , wheelbase_(wheelbase)
    , rear_length_(rear_length)
{
    config_.validate();
}

void LowSpeedSafety::reset()
{
    engaged_ = false;
}

void LowSpeedSafety::apply(std::vector<double>& state, double speed, bool update_latch)
{
    if (update_latch) {
        if (engaged_) {
            if (speed > config_.release_speed) {
                engaged_ = false;
            }
        } else {
            if (speed < config_.engage_speed) {
                engaged_ = true;
            }
        }
    }

    const auto yaw_target = kinematic_yaw_rate(state, speed);
    const auto slip_target = kinematic_slip(state, speed);
    const auto lateral_target = kinematic_lateral_velocity(state, speed);

    if (longitudinal_index_ && index_in_bounds(*longitudinal_index_, state)) {
        const std::size_t idx = static_cast<std::size_t>(*longitudinal_index_);
        double value = state[idx];
        if (value < 0.0) {
            state[idx] = 0.0;
        } else if (!engaged_ && std::fabs(value) <= config_.stop_speed_epsilon) {
            state[idx] = 0.0;
        }
    }

    if (yaw_rate_index_ && index_in_bounds(*yaw_rate_index_, state)) {
        const std::size_t idx = static_cast<std::size_t>(*yaw_rate_index_);
        if (engaged_) {
            const double limit = config_.yaw_rate_limit;
            const double target = yaw_target.value_or(0.0);
            state[idx] = clamp(target, -limit, limit);
        } else {
            const double limit = config_.yaw_rate_limit;
            state[idx] = clamp(state[idx], -limit, limit);
        }
    }

    if (lateral_index_ && index_in_bounds(*lateral_index_, state)) {
        const std::size_t idx = static_cast<std::size_t>(*lateral_index_);
        double value = state[idx];
        if (engaged_) {
            if (lateral_target.has_value()) {
                state[idx] = *lateral_target;
            } else {
                const double limit = config_.stop_speed_epsilon;
                state[idx] = clamp(value, -limit, limit);
            }
        } else if (std::fabs(value) <= config_.stop_speed_epsilon) {
            state[idx] = 0.0;
        }
    }

    if (slip_index_ && index_in_bounds(*slip_index_, state)) {
        const std::size_t idx = static_cast<std::size_t>(*slip_index_);
        if (engaged_) {
            const double limit = config_.slip_angle_limit;
            const double target = slip_target.value_or(0.0);
            state[idx] = clamp(target, -limit, limit);
        } else {
            const double limit = config_.slip_angle_limit;
            state[idx] = clamp(state[idx], -limit, limit);
        }
    }

    if (!wheel_speed_indices_.empty()) {
        for (int raw_idx : wheel_speed_indices_) {
            if (!index_in_bounds(raw_idx, state)) {
                continue;
            }
            const std::size_t idx = static_cast<std::size_t>(raw_idx);
            double value = state[idx];
            if (value <= 0.0) {
                state[idx] = 0.0;
            } else if (engaged_ && value <= config_.stop_speed_epsilon) {
                state[idx] = 0.0;
            }
        }
    }
}

bool LowSpeedSafety::index_in_bounds(int index, const std::vector<double>& state)
{
    return index >= 0 && static_cast<std::size_t>(index) < state.size();
}

double LowSpeedSafety::clamp(double value, double min_value, double max_value)
{
    return std::max(min_value, std::min(max_value, value));
}

std::optional<double> LowSpeedSafety::kinematic_beta(const std::vector<double>& state) const
{
    if (!steering_index_.has_value() || !wheelbase_.has_value() || !rear_length_.has_value()) {
        return std::nullopt;
    }
    if (!index_in_bounds(*steering_index_, state)) {
        return std::nullopt;
    }
    if (*wheelbase_ <= 0.0) {
        return std::nullopt;
    }
    const double delta = state[static_cast<std::size_t>(*steering_index_)];
    const double ratio = *rear_length_ / *wheelbase_;
    return std::atan(std::tan(delta) * ratio);
}

std::optional<double> LowSpeedSafety::kinematic_yaw_rate(const std::vector<double>& state, double speed) const
{
    const auto beta = kinematic_beta(state);
    if (!beta.has_value() || !steering_index_.has_value() || !wheelbase_.has_value()) {
        return std::nullopt;
    }
    if (!index_in_bounds(*steering_index_, state)) {
        return std::nullopt;
    }
    if (std::fabs(speed) <= 1e-9) {
        return 0.0;
    }
    const double delta = state[static_cast<std::size_t>(*steering_index_)];
    if (*wheelbase_ <= 0.0) {
        return std::nullopt;
    }
    return speed * std::cos(*beta) * std::tan(delta) / *wheelbase_;
}

std::optional<double> LowSpeedSafety::kinematic_slip(const std::vector<double>& state, double)
    const
{
    (void)state;
    return kinematic_beta(state);
}

std::optional<double> LowSpeedSafety::kinematic_lateral_velocity(const std::vector<double>& state, double speed) const
{
    const auto beta = kinematic_beta(state);
    if (!beta.has_value()) {
        return std::nullopt;
    }
    return speed * std::sin(*beta);
}

} // namespace vehiclemodels::sim
